pkg/scan/docker/docker.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/mholt/archiver/v3"
+	"github.com/palantir/log4j-sniffer/pkg/archive"
+	"github.com/palantir/log4j-sniffer/pkg/crawl"
+	"github.com/palantir/log4j-sniffer/pkg/scan"
+	"github.com/pkg/errors"
+)
+
+type Scanner struct {
+	config     scan.Config
+	crawler    crawl.Crawler
+	reporter   *crawl.Reporter
+	identifier crawl.Identifier
+	client     client.CommonAPIClient
+}
+
+func NewDockerScanner(config scan.Config, stdout, stderr io.Writer) (*Scanner, error) {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create docker client")
+	}
+
+	return &Scanner{
+		config: config,
+		crawler: crawl.Crawler{
+			ErrorWriter: stderr,
+			IgnoreDirs:  config.Ignores,
+		},
+		reporter: &crawl.Reporter{
+			OutputJSON:      config.OutputJSON,
+			OutputWriter:    stdout,
+			DisableCVE45105: config.DisableCVE45105,
+		},
+		identifier: crawl.NewIdentifier(config.ArchiveListTimeout, archive.WalkZipFiles, archive.WalkTarGzFiles),
+		client:     dockerClient,
+	}, nil
```

Comment:
This seems pretty similar to the other scanner for the files on disk.

Is there a way we can have a reusable scanner and then provide it with some dependency that's an abstraction reusable across both filesystem and/or docker? Then we get benefit of only ever having 1 way to wire up a scanner.

Reply:
Have refactored things around to use a new `NewScannerFromConfig` function which will initialize a scanner with a configured crawler, identifier and reporter, updated both the filesystem and docker scan/crawl functions to use this